// @skyware/bot/dist/struct/post/embed/EmbedImage.js
/**
 * Represents an image embedded in a post.
 */
export class EmbedImage {
    /** The image's CID. */
    cid;
    /** The image's alt text. */
    alt;
    /** The image's MIME type. */
    mimeType;
    /** Image size in bytes. */
    size;
    /** The image's aspect ratio. */
    aspectRatio;
    /** The URL of the image's thumbnail located at the Bluesky CDN. */
    _thumbUrl;
    /** The URL of the full-size image located at the Bluesky CDN. */
    _fullsizeUrl;
    /**
     * @param data The image data.
     */
    constructor({ cid, alt = "", mimeType, size, aspectRatio, thumb, fullsize }) {
        this.cid = cid;
        this.alt = alt;
        this.mimeType = mimeType;
        this.size = size;
        if (aspectRatio)
            this.aspectRatio = aspectRatio;
        if (thumb)
            this._thumbUrl = thumb;
        if (fullsize)
            this._fullsizeUrl = fullsize;
    }
    /**
     * Get the URL of the image's thumbnail.
     */
    get thumbnailUrl() {
        return this._thumbUrl;
    }
    /**
     * Get the URL of the full-size image.
     */
    get url() {
        return this._fullsizeUrl;
    }
}
