// @skyware/bot/dist/struct/post/embed/VideoEmbed.js
import { PostEmbed } from "./PostEmbed.js";
/**
 * A post embed that contains a video.
 */
export class VideoEmbed extends PostEmbed {
    /** The video's CID. */
    cid;
    /** A link to the [master playlist file](https://jazco.dev/2024/07/05/hls#master-playlists) for the video. */
    playlist;
    /** The video's thumbnail. */
    thumbnail;
    /** The video's alt text. */
    alt;
    /** The video's aspect ratio. */
    aspectRatio;
    /** Any captions associated with the video. */
    captions;
    /**
     * @param data Embed data.
     */
    constructor({ cid, playlist, thumbnail, alt, aspectRatio, captions }) {
        super();
        this.cid = cid;
        this.playlist = playlist;
        if (thumbnail)
            this.thumbnail = thumbnail;
        if (alt)
            this.alt = alt;
        if (aspectRatio)
            this.aspectRatio = aspectRatio;
        if (captions)
            this.captions = captions;
    }
    isVideo() {
        return true;
    }
    /**
     * Constructs a VideoEmbed from an embed view and a record.
     * @param videoView The view of the embed.
     * @param videoRecord The embed video record.
     */
    static fromView(videoView, videoRecord) {
        return new VideoEmbed({
            cid: videoView.cid,
            playlist: videoView.playlist,
            thumbnail: videoView.thumbnail,
            alt: videoView.alt,
            aspectRatio: videoView.aspectRatio,
            captions: videoRecord.captions,
        });
    }
}
